<a href="https://colab.research.google.com/github/patel7147/ROYAL_DL/blob/master/Iris_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
from keras.utils import to_categorical

Iris Dataset is a dataset consits of plants sample 

In [2]:
dataset = pd.read_csv('/content/Iris.csv')
dataset

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

In [20]:
x = dataset.iloc[:,1:5].values
y = dataset.iloc[:,5].values

from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
y1 = encoder.fit_transform(y)
Y = pd.get_dummies(y1).values
#print(Y)

In [21]:
from sklearn.model_selection import train_test_split

x_train,x_test,y1_train,y1_test = train_test_split(x,Y,test_size=0.2)
print(x_test)

[[5.1 3.5 1.4 0.2]
 [6.3 2.5 5.  1.9]
 [4.6 3.6 1.  0.2]
 [5.1 3.4 1.5 0.2]
 [5.8 2.8 5.1 2.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.  4.2 1.2]
 [5.3 3.7 1.5 0.2]
 [5.5 2.4 3.8 1.1]
 [4.7 3.2 1.6 0.2]
 [6.3 2.7 4.9 1.8]
 [6.2 2.9 4.3 1.3]
 [5.4 3.9 1.3 0.4]
 [4.8 3.1 1.6 0.2]
 [7.2 3.  5.8 1.6]
 [6.4 3.2 5.3 2.3]
 [4.8 3.4 1.9 0.2]
 [6.8 3.  5.5 2.1]
 [6.1 2.6 5.6 1.4]
 [6.  2.2 5.  1.5]
 [5.7 2.5 5.  2. ]
 [5.6 3.  4.5 1.5]
 [4.3 3.  1.1 0.1]
 [4.9 3.1 1.5 0.1]
 [5.4 3.4 1.5 0.4]
 [4.9 3.  1.4 0.2]
 [6.9 3.1 4.9 1.5]
 [5.4 3.7 1.5 0.2]
 [6.1 2.8 4.  1.3]
 [6.5 3.2 5.1 2. ]]


Here we split the data in the test_Data and train_data 


In [83]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD, Adam

model = Sequential()

model.add(Dense(10, input_shape=(4,),activation='tanh'))
model.add(Dense(8,activation='tanh'))
model.add(Dense(6,activation='tanh'))
model.add(Dense(3,activation='softmax'))

model.compile(Adam(lr=1),'categorical_crossentropy',metrics=['accuracy'])

model.summary()

Model: "sequential_36"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_144 (Dense)            (None, 10)                50        
_________________________________________________________________
dense_145 (Dense)            (None, 8)                 88        
_________________________________________________________________
dense_146 (Dense)            (None, 6)                 54        
_________________________________________________________________
dense_147 (Dense)            (None, 3)                 21        
Total params: 213
Trainable params: 213
Non-trainable params: 0
_________________________________________________________________


In [84]:
model.fit(x_train,y1_train,epochs=200)
y_pred = model.predict(x_test)

Epoch 1/200
4/4 [==============================] - 0s 2ms/step - loss: 2.7628 - accuracy: 0.4250
Epoch 2/200
4/4 [==============================] - 0s 2ms/step - loss: 4.6795 - accuracy: 0.3750
Epoch 3/200
4/4 [==============================] - 0s 4ms/step - loss: 4.5535 - accuracy: 0.3333
Epoch 4/200
4/4 [==============================] - 0s 3ms/step - loss: 2.9008 - accuracy: 0.2833
Epoch 5/200
4/4 [==============================] - 0s 3ms/step - loss: 2.2262 - accuracy: 0.3083
Epoch 6/200
4/4 [==============================] - 0s 3ms/step - loss: 1.5740 - accuracy: 0.3333
Epoch 7/200
4/4 [==============================] - 0s 2ms/step - loss: 1.7748 - accuracy: 0.3333
Epoch 8/200
4/4 [==============================] - 0s 2ms/step - loss: 1.6514 - accuracy: 0.2750
Epoch 9/200
4/4 [==============================] - 0s 2ms/step - loss: 1.5170 - accuracy: 0.3250
Epoch 10/200
4/4 [==============================] - 0s 2ms/step - loss: 1.2306 - accuracy: 0.3417
Epoch 11/200
4/4 [===========

**Using SGD(Stochastic Gradient Descent) Optimizer with Loss function Mean_Squared_Error:**


---



1)Learning_Rate=0.01,Total parameters: 213,Epochs=200,
Then- 
Loss: 0.1333 , 
Accuracy: 0.6667

2)Learning_Rate=0.04,Total parameters: 213,Epochs=200,
Then-
Loss: 0.0308 , Accuracy: 0.9750

3)Learning_Rate=0.06,Total parameters: 213,Epochs=200,
Then-
Loss: 0.0308 , Accuracy: 0.9750

4)Learning_Rate=1,Total parameters: 213,Epochs=200,
Then-
Loss: 0.1032 , Accuracy: 0.8083

=>Here, learning_rate is taken as a variable by keeping other constants, In this example , the error came same till learning_rate=0.08, after 0.08 the accuracy gets less then before. 

=>The accuracy doesn't change even epochs=1000 at learning rate=0.06, so the max accuracy in the mean_squared_Error is 0.9750


**Using SGD(Stochastic Gradient Descent) Optimizer with Loss function Categorical_CrossEntropy:**


---


1)Learning_Rate=0.01,Total parameters: 213,Epochs=200,
Then- 
Loss: 0.4908 , Accuracy: 0.6667

2)Learning_Rate=0.04,Total parameters: 213,Epochs=200,
Then- 
 Loss: 0.1026 , Accuracy: 0.9583

3)Learning_Rate=0.06,Total parameters: 213,Epochs=200,
Then-
Loss: 0.0975 , Accuracy: 0.9833

4)Learning_Rate=1,Total parameters: 213,Epochs=200,
Then-
Loss: 1.1071 , Accuracy: 0.3333

=>Here , the learning rate is taken as variable, the heighest accuracy is measured at learning rate=0.06 after that the accuracy gets lower then before.

=>Here, after taking epochs=1000 the accuracy does'nt change so the max accuracy we get in cross_Entropy loss function is:0.9833


**Using Adam(Adaptive Moment Estimation) Optimizer with Loss function Mean_Squared_Error:**

---



1)Learning_Rate=0.01,Total parameters: 213,Epochs=200, 
Then-
Loss: 0.0100 , Accuracy: 0.9750

2)Learning_Rate=0.04,Total parameters: 213,Epochs=200, 
Then-
Loss: 0.0121 , Accuracy: 0.9833

3)Learning_Rate=0.06,Total parameters: 213,Epochs=200, 
Then-
Loss: 0.0200 , Accuracy: 0.9500

4)Learning_Rate=1,Total parameters: 213,Epochs=200, 
Then-
loss: 0.4444 - accuracy: 0.3333

=>Here, the max accuracy gets at epochs=200 , and learning rate =0.04. The learning rate is variable other is constant.

=>When the epochs=1000, the accuracy gets max at learning_rate=0.01, the accuracy=0.9750.





**Using Adam(Adaptive Moment Estimation) Optimizer with Loss function Categorical_CrossEntropy:**


---

1)Learning_Rate=0.01,Total parameters: 213,Epochs=200, 
Then-
Loss: 0.0233 , Accuracy: 0.9833

2)Learning_Rate=0.04,Total parameters: 213,Epochs=200, 
Then-Loss: 0.0741 , Accuracy: 0.9667

3)Learning_Rate=0.06,Total parameters: 213,Epochs=200, 
Then-Loss: 0.0861 , Accuracy: 0.9667

4)Learning_Rate=0.06,Total parameters: 213,Epochs=200, 
Then-Loss: 1.2075 , Accuracy: 0.2500

=>Here, the max accuracy gets at epochs=200 , and learning rate =0.01. The learning rate is variable other is constant. After , 0.01 the accuracy gets lower than the previous one.

=>When the epochs=1000, the accuracy gets max at learning_rate=0.01, the accuracy=0.9833.

**Conclusion:**

---
*Here we have taken two different optimizers: 
1)SGD optimizer
2)Adam Optimizer

->In each optimizer we have taken two loss function:
1)Mean_Squared_Error
2)Categorical_CrossEntropy

=>Using SGD optimizer, the highest accuracy we get in mean_squared_error:
>Learning_Rate=0.04,Total parameters: 213,Epochs=200, Then- Loss: 0.0308 , Accuracy: 0.9750

And in Categorical_CrossEntropy is:
>Learning_Rate=0.06,Total parameters: 213,Epochs=200, Then- Loss: 0.0975 , Accuracy: 0.9833 

=>Using Adam optimizer, the highest accuracy we get in mean_squared_error:
>Learning_Rate=0.04,Total parameters: 213,Epochs=200, Then- Loss: 0.0121 , Accuracy: 0.9833

And in Categorical_CrossEntropy is:
>Learning_Rate=0.01,Total parameters: 213,Epochs=200, Then- Loss: 0.0233 , Accuracy: 0.9833